### PR TITLE
Working Kubernetes branch

### PR DESCRIPTION
**Pull request by MatMeredith**
Peter.   I'd like to merge this into master.   Could I ask you to review it please?   It's basically your branch with a set of changes that I made to get working calls on GKE (and in some places to remove references to stuff outside of PC etc. and made one or two simplifications like defaulting to a 1 pod etcd cluster).   Questions...
- I assume the changes to icscf_uri, scscf_uri  etc. in the base Dockerfile were so that we can have just sprout in  DNS rather than icscf.sprout and scscf.sprout?   Would it have been difficult to get Kubernetes to support both?
- I've added "loadbalancer" config to the Bono and Ellis yaml files so that the services are externally accessible.   This obviously depends on Kubernetes running in a cloud provider that supports it but I didn't see any other obvious default...   I've commented in the readme that it might need to be changed.

### CODE REVIEW

File: base/Dockerfile [hunk 1]
```diff
@@ -17,9 +18,27 @@ COPY sysctl /sbin/sysctl
...
 RUN /etc/init.d/clearwater-auto-config-docker restart
 RUN /etc/init.d/clearwater-infrastructure restart
 COPY clearwater-infrastructure.supervisord.conf /etc/supervisor/conf.d/clearwater-infrastructure.conf
 COPY clearwater-group.supervisord.conf /etc/supervisor/conf.d/clearwater-group.conf
+COPY snmpd.supervisord.conf /etc/supervisor/conf.d/snmpd.conf
+
+# Changes to make everything work with Kubernetes
+RUN apt-get install -y netcat-openbsd
+RUN sed -i 's/SCTP_streams.*/No_SCTP;/' /usr/share/clearwater/bin/generic_create_diameterconf
+RUN sed -i 's/scscf.sprout/sprout/' /etc/init.d/clearwater-auto-config-docker
```
Comment (resouer) [inline base/Dockerfile]: Emm, I can confirm `sed -i 's/scscf.sprout/sprout/'` works.

File: README.md [hunk 2]
```diff
@@ -61,13 +68,45 @@ Note that scaling of Docker deployments is a work in progress and there are curr
...
+Instead of using Docker Compose, you can deploy Clearwater in Kubernetes. This
+requires a Kubernetes cluster, and a Docker repository.
+
+### Prepare the images and edit the yaml files
+
+- First, build all the required images locally.
+
+        # Build the Clearwater docker images.
+        cd clearwater-docker
+        for i in base memcached cassandra chronos bono ellis homer homestead ralf sprout ; do docker build -t clearwater/$i $i ; done
```
Comment (resouer) [inline README.md]: Is `memcached` still available?

File: README.md [hunk 3]
```diff
@@ -61,13 +68,45 @@ Note that scaling of Docker deployments is a work in progress and there are curr
...
+        do
+            docker tag -f clearwater/$i:latest path_to_your_repo/clearwater/$i:latest
+            docker push path_to_your_repo/clearwater/$i:latest
+        done
+
+- Update the Kubernetes yaml to match your deployment.   
+  - In each file ending depl.yaml you will need to:
+    - edit the image path to match the path to the repository that you pushed your images to
+    - edit the value of the ZONE attribute to match the domain of your Kubernetes cluster -- by default it is "default.svc.cluster.local" which will work for a "default" Kubernetes cluster
+  - The bono-svc.yaml and ellis-svc.yaml files configure external load balancers to expose SIP access and provisioning access from outside of the cluster.   This will work as is if you are running on a cloud provider that support this -- e.g. GKE -- but alternative configuration will be required for Kubernetes running in other environments.
```
Comment (resouer) [inline README.md]: `type: "NodePort"` can be used here.

File: README.md [hunk 4]
```diff
@@ -52,22 +59,54 @@ To start the Clearwater services, run:
...
+Instead of using Docker Compose, you can deploy Clearwater in Kubernetes. This
+requires a Kubernetes cluster, and a Docker repository.
+
+### Prepare the images and edit the yaml files
+
+- First, build all the required images locally.
+
+        # Build the Clearwater docker images.
+        cd clearwater-docker
+        for i in base astaire cassandra chronos bono ellis homer homestead ralf sprout ; do docker build -t clearwater/$i $i ; done
```
Comment (resouer) [inline README.md:89]: After this step, I noticed there's no `/etc/cassandra/cassandra.yaml ` in the cassandra image, is it expected?

Comment (MatMeredith) [reply README.md:89]: Yes, that's right.   /etc/cassandra/cassandra.yaml gets created automatically by the [Cassandra plugin](https://github.com/Metaswitch/clearwater-etcd-plugins/blob/e41f067148f7e2de0409ada8d8b028c6853fc2a6/clearwater_cassandra/cassandra_plugin.py) to clearwater cluster manager.

Comment (resouer) [reply README.md:89]: While during testing cassandra keeps reporting:
```
java.rmi.server.ExportException: Port already in use: 7199; nested exception is: 
```
I am sure `7199` is free, so I have to add `$JVM_OPTS -Dcom.sun.management.jmxremote.ssl=true` according to [this work around](http://stackoverflow.com/questions/33725162/cassandra-jmx-java-rmi-server-exportexception-port-already-in-use-7199)

Any idea about this?

Comment (MatMeredith) [reply README.md:89]: Sorry -- I've not seen this before either running directly under Docker, or running under Kubernetes on GKE.   What networking model are you using?   Is it possible that 7199 is already open on the container host?

Comment (resouer) [reply README.md:89]: 7199 port error disappeared somehow ...  Now the only problem I can see is homestead:
```
20-04-2017 13:27:32.102 UTC Status load_monitor.cpp:105: Constructing LoadMonitor
20-04-2017 13:27:32.102 UTC Status load_monitor.cpp:106:    Target latency (usecs)   : 100000
20-04-2017 13:27:32.102 UTC Status load_monitor.cpp:107:    Max bucket size          : 1000
20-04-2017 13:27:32.102 UTC Status load_monitor.cpp:108:    Initial token fill rate/s: 100.000000
20-04-2017 13:27:32.102 UTC Status load_monitor.cpp:109:    Min token fill rate/s    : 10.000000
20-04-2017 13:27:32.102 UTC Status dnscachedresolver.cpp:150: Creating Cached Resolver using servers:
20-04-2017 13:27:32.102 UTC Status dnscachedresolver.cpp:160:     10.96.0.10
20-04-2017 13:27:32.102 UTC Status a_record_resolver.cpp:54: Created ARecordResolver
20-04-2017 13:27:32.102 UTC Status a_record_resolver.cpp:54: Created ARecordResolver
20-04-2017 13:27:32.102 UTC Status cassandra_store.cpp:181: Configuring store connection
20-04-2017 13:27:32.102 UTC Status cassandra_store.cpp:182:   Hostname:  cassandra.default.svc.cluster.local
20-04-2017 13:27:32.103 UTC Status cassandra_store.cpp:183:   Port:      9160
20-04-2017 13:27:32.103 UTC Status cassandra_store.cpp:211: Configuring store worker pool
20-04-2017 13:27:32.103 UTC Status cassandra_store.cpp:212:   Threads:   10
20-04-2017 13:27:32.103 UTC Status cassandra_store.cpp:213:   Max Queue: 0
20-04-2017 13:27:32.114 UTC Error main.cpp:756: Failed to initialize the Cassandra cache with error code 1.
20-04-2017 13:27:32.114 UTC Status main.cpp:757: Homestead is shutting down
```

Not sure about the reason.

File: ellis/Dockerfile [hunk 5]
```diff
@@ -5,6 +5,7 @@ RUN apt-get update && DEBIAN_FRONTEND=noninteractive apt-get install -y --force-
 RUN /etc/init.d/mysql start && apt-get update && DEBIAN_FRONTEND=noninteractive apt-get install -y --force-yes ellis
 RUN /etc/init.d/mysql start
 
+COPY create_numbers.sh /usr/share/clearwater/ellis/create_numbers.sh
```
Comment (resouer) [inline ellis/Dockerfile]: Latest ellis is trying to find `http://hs.default.svc.cluster.local:8889`, which I think should be `http://headstead.default.svc.cluster.local:8889`

File: README.md [hunk 6]
```diff
@@ -52,22 +59,62 @@ To start the Clearwater services, run:
...
+
+- Next, push them to your repository (which must be accessible from the Kubernetes deployment)
+
+        for i in base astaire cassandra chronos bono ellis homer homestead ralf sprout
+        do
+            docker tag -f clearwater/$i:latest path_to_your_repo/clearwater/$i:latest
```
Comment (resouer) [inline README.md]: Seems no `docker tag -f`, just docker tag is enough

**Review by plwhite — COMMENTED** [review]
Looks like the gitignore file either is or was failing to tidy out ".pyc" binary files. Probably worth fixing that one up.

**Comment by plwhite** [comment]
I had a look through. Nothing except the DNS leaps out as being problematic, and if it works for you I think you can mostly release it - I'll make sure we try it out in due course in case of anything odd in an environmental sense.

The problem with DNS was that Kubernetes already has a DNS namespacing structure, which means that DNS names are all something like: `svc.ns.base` where `svc` is the service name, `ns` is the namespace name, and `base` is some global thing such as `site1.kubernetes.local`. Each pod has a base DNS of `ns.base`. You cannot put dots into either service or namespace names to avoid breaking this, and so you cannot have a DNS lookup of "iscf.sprout". There may be some way to get round this (other than just changing the name used - you could play games with a local dnsmasq instance, for example), but it's annoyingly deeply baked in.

**Comment by resouer** [comment]
@plwhite where is `iscf.sprout` used?

Kubernetes can set subdomain, for example: `myapp.mysubdomain`, does that help?

**Comment by resouer** [comment]
@MatMeredith Here's an example:

```
apiVersion: v1
kind: Service
metadata:
  name: sprout
spec:
  selector:
    service: sprout
  clusterIP: None
  ports:
    - name: foo # Actually, no port is needed.
      port: 1234 
      targetPort: 1234
```
Then two `sprout` deploys, one for `iscf`, one for `scscf`

```
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: iscf-sprout
spec:
  replicas: 1
  selector:
    matchLabels:
      service: sprout
  template:
    metadata:
      labels:
        service: sprout
        snmp: enabled
    spec:
      hostname: iscf
      subdomain: sprout
      containers:
      - image: busybox
        command:
        - sleep
        - "3600"
        name: busybox
```

```
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: scscf-sprout
spec:
  replicas: 1
  selector:
    matchLabels:
      service: sprout
  template:
    metadata:
      labels:
        service: sprout
        snmp: enabled
    spec:
      hostname: scscf
      subdomain: sprout
      containers:
      - image: busybox
        command:
        - sleep
        - "3600"
        name: busybox
```
@plwhite I can make sure DNS part works in my example, any other problem in your mind?

File: README.md (hunk 4 above)
Comment (MatMeredith) [reply README.md:89]: If you spin up all the components at the same time then Homestead usually restarts a few times like this whilst it waits for Cassandra to start up.   Homestead can't run without Cassandra, and Cassandra takes a little time to come up...   Is that not what you are seeing?  If not, I'd check that your Cassandra pod is healthy and that Homestead can reach it.

**Comment by MatMeredith** [comment]
@plwhite Thanks.   Yes, I'd figured the issue you were working around with the DNS and I think I've maintained your behaviour, but I've moved the functionality to https://github.com/Metaswitch/clearwater-infrastructure/pull/441 instead of having these overrides in the Dockerfile.

@resouer So, the way that Clearwater usually works is described here: http://clearwater.readthedocs.io/en/stable/Clearwater_DNS_Usage.html.   Specifically the Sprout cluster should be accessible via either icscf.sprout.<zone> or scscf.sprout.<zone> (and ideally SRV lookups on these names would return port 5052 or 5054 respectively).

I think that means that we need the an ICSCF sprout service exposed as icscf.sprout.<zone> and mapping to port 5052 on the underlying pods and an SCSCF sprout service exposed as scscf.sprout.<zone> and mapping to port 5054, but it looks like your subdomains are being applied to the pods?   Or am i misunderstanding?  Is there any way to achieve what is required here?

**Comment by resouer** [comment]
@MatMeredith It seems https://github.com/Metaswitch/clearwater-infrastructure/pull/441/files already fix the dns problem above?

Yes my yaml is only a workaround and subdomain is `pod-hostname.subdomain`

**Comment by MatMeredith** [comment]
@resouer Yes, that PR "solves" the issue, but it's really only a workaround rather than a fix.

**Pull request merged**